Proposal

PolyProgrammist opened the pull request.

PolyProgrammist commented:
deployed on testnet here: proposal_feature.testnet
sample commands here: https://github.com/NEAR-DevHub/neardevhub-contract/issues/89#issuecomment-1924182487

frol commented on src/lib.rs:
```diff
@@ -637,13 +878,46 @@ mod tests {
             .build()
     }
 
+    #[allow(dead_code)]
```
Remove unused code

PolyProgrammist replied:
It is used by the other test files. But for some reason rust thinks they are not used at all

PolyProgrammist commented:
> > deployed on testnet here: proposal_feature.testnet sample commands here: [#89 (comment)](https://github.com/NEAR-DevHub/neardevhub-contract/issues/89#issuecomment-1924182487)
> 
> All these sample calls are also in the tests, right? So that we can also use the test as an example of how to call it?

Yes, in tests/proposal.rs

ailisp commented on src/lib.rs:
```diff
@@ -198,8 +229,81 @@ impl Contract {
+        let proposal = Proposal {
+            id,
+            author_id: author_id.clone(),
+            social_db_post_block_height: 0u64, // TODO
```
```suggestion
            social_db_post_block_height: 0u64,
```

ailisp reviewed (dismissed):
Great work, solid implementation and extensive tests! I don't think it will break anything in production given proposals are added as separate fields and migration is done properly. Please solve the merge conflict and I think it's ready to merge.

PolyProgrammist committed:
New Proposal workflow

PolyProgrammist commented:
Force pushed as rebase needed. Previous version can be found here https://github.com/PolyProgrammist/neardevhub-contract/pull/3

frol commented on src/lib.rs:
```diff
@@ -199,8 +231,81 @@ impl Contract {
+            snapshot_history: vec![],
+        };
+
+        proposal::repost::publish_to_socialdb_feed(
```
Return this `Promise` as a result of `add_proposal` call. See https://github.com/NEAR-DevHub/neardevhub-contract/pull/88#discussion_r1476879100 for more details about returning a Promise

frol commented on src/lib.rs:
```diff
@@ -376,6 +551,99 @@ impl Contract {
+        require!(
+            self.has_moderator(editor_id.clone())
+                || editor_id.clone() == env::current_account_id()
+                || (is_draft(&proposal.snapshot.body.clone().latest_version().timeline)
+                    && (is_empty_review(&proposal_body.timeline)
+                        || is_draft(&proposal_body.timeline))),
```
It is worth implementing these functions as methods of Timeline enum instead of free-form functions.

frol commented on src/proposal/repost.rs:
```diff
@@ -0,0 +1,43 @@
+use crate::{get_proposal_description, social_db::social_db_contract, Proposal};
+
+fn repost_internal(proposal: Proposal, contract_address: AccountId) -> near_sdk::serde_json::Value {
+    let proposal_link = format!("/devgovgigs.near/widget/Proposal?id={}", proposal.id);
```
@Megha-Dev-19 What should be the link to the Proposal? `/devhub.near/widget/app?page=proposal&id={}`?

Megha-Dev-19 replied:
Yes that’s correct 👍

frol commented on src/proposal/mod.rs:
```diff
@@ -0,0 +1,151 @@
+use near_sdk::serde::{Deserialize, Serialize};
+use near_sdk::{AccountId, BlockHeight, Timestamp};
+
+pub type ProposalId = u64;
```
Let's either use u32 or near_sdk::json_type::U64, so we don't cause troubles for JS consumers that can only handle number values in JSON up to 2**53 (f64)

frol commented on src/proposal/mod.rs:
```diff
@@ -0,0 +1,151 @@
+        serialize_with = "u128_dec_format::serialize",
+        deserialize_with = "u128_dec_format::deserialize"
+    )]
+    pub requested_sponsorship_amount: Balance,
```
Balance u128 is not going to play well with JSON serialization. Given that we only deal with integer amounts (without fractional part), let's just use u32 for the balance in proposals (try moving the Balance definition from lib.rs to post/mod.rs if that won't trigger changes outside of post/ module, so we can kill it together with the post module when the time will come).

PolyProgrammist committed:
Review:
Return Promise for add_proposal
Methods for timeline
ProposalId to u32
Replaced Balance in proposal for u32, moved Balance from lib.rs to post/mod.rs

PolyProgrammist commented:
Squashed small additional change into Review commit

frol commented on src/proposal/repost.rs:
```diff
@@ -0,0 +1,43 @@
+    let desc = get_proposal_description(proposal.snapshot.body.clone());
+
+    let text = format!(
+        "@{author} [Created the proposal on DevHub]({post_link})\n{title}{desc}",
```
This formatting is misleading when we open it in the feed: https://near.org/near/widget/ProfilePage?accountId=devgovgigs.near

<img width="937" alt="image" src="https://github.com/NEAR-DevHub/neardevhub-contract/assets/304265/e9ad4077-7d4f-41b1-99b1-48fc90101468">

It looks like DevHub endorses the content unless you read really carefully.

I suggest using the following text:

```suggestion
        "There is a new proposal on DevHub from @{author}: “{title}“\n> {summary}\n__Read the full proposal [here]({post_link})__”,
```

Use `summary` instead of the `description` of the proposal.

frol commented on src/lib.rs:
```diff
@@ -123,6 +136,23 @@ impl Contract {
+
+    pub fn get_all_proposal_ids(&self) -> Vec<ProposalId> {
+        near_sdk::log!("get_all_proposal_ids");
+        (0..self.proposals.len()).map(|x| x.try_into().unwrap()).collect()
```
```suggestion
        (0..self.proposals.len().try_into().unwrap()).collect()
```

frol commented on src/lib.rs:
```diff
@@ -199,8 +229,83 @@ impl Contract {
+    #[payable]
+    pub fn add_proposal(&mut self, body: VersionedProposalBody, labels: HashSet<String>) -> Promise {
+        near_sdk::log!("add_proposal");
+        let id = self.proposals.len();
```
Let's convert it once here and don't `try_into()` it below.

```suggestion
        let id: ProposalId = self.proposals.len().try_into().unwrap();
```

frol commented on src/lib.rs:
```diff
@@ -621,13 +889,20 @@ impl Contract {
+#[serde(crate = "near_sdk::serde")]
+#[schemars(crate = "near_sdk::schemars")]
+pub struct BlockHeightCallbackRetValue {
+    proposal_id: near_sdk::json_types::U64,
```
It should be ProposalId here, shouldn't it?

frol commented on src/lib.rs:
```diff
@@ -376,6 +551,99 @@ impl Contract {
+        );
+        let editor_id = env::predecessor_account_id();
+        let mut proposal: Proposal =
+            self.proposals.get(id.try_into().unwrap()).unwrap_or_else(|| panic!("Proposal id {} not found", id)).into();
```
I don't think we need `try_into` here since `id` is already of type `ProposalId`

```suggestion
            self.proposals.get(id).unwrap_or_else(|| panic!("Proposal id {} not found", id)).into();
```

PolyProgrammist replied:
Here id is an index of a vector. And it should be u64. But ProposalId is u32. So I would change it for `id.into()`

frol commented on src/post/mod.rs:
```diff
@@ -56,21 +59,26 @@ pub enum StorageKey {
     LabelToPostsV2,
     AuthorToAuthorPosts,
     AuthorPosts(CryptoHash),
+    Proposals,
+    LabelToProposals,
+    AuthorProposals,
```
Oh, you have to add new keys to the end of the list, since the order of the keys is used to assign borsh-serialized value. NEAR SDK collections operate on the contract "local storage" and these keys are used as a prefix, so since there is already data stored with the prefixes 12-13 (Communities, AddOns), we don't want to break it by moving Communities to have prefix 15 since there is already data stored under prefix 12.

PolyProgrammist committed:
Review:
Avoid try_into where possible
Use ProposalId in BlockHeightCallbackRetValue
Change link for socialdb
Fixed StorageKey enum

PolyProgrammist committed:
Fix link for post

frol commented on src/proposal/mod.rs:
```diff
@@ -0,0 +1,151 @@
+    }
+}
+
+pub fn get_proposal_summary(proposal_body: VersionedProposalBody) -> String {
+    proposal_body.clone().latest_version().summary
+}
```
It seems that this function is only used in a single place, so I would inline it there.

frol commented on src/lib.rs:
```diff
@@ -376,6 +551,99 @@ impl Contract {
+        );
+        let editor_id = env::predecessor_account_id();
+        let mut proposal: Proposal =
+            self.proposals.get(id.into()).unwrap_or_else(|| panic!("Proposal id {} not found", id)).into();
```
Are these intos necessary?

```suggestion
            self.proposals.get(id).unwrap_or_else(|| panic!("Proposal id {} not found", id));
```

PolyProgrammist replied:
Yes, the for first one an `id` is an index of a vector. And it should be u64. But ProposalId is u32
For the second one, it is casting from `VersionedProposal` to `Proposal`

frol reviewed (dismissed):
@PolyProgrammist I am approving it and overall, ready to merge it (will do it tomorrow morning since it is Friday evening...)

PolyProgrammist committed:
Fix tests for post link

PolyProgrammist committed:
Review: inline get_proposal_summary

PolyProgrammist committed:
Remove get_proposal_summary
